Checkout/Checkout/Test/Validation/CVVValidatorTests.swift

```diff
@@ -50,89 +50,38 @@ final class CVVValidatorTests: XCTestCase {
     }
   }
 
-  func test_validateCVV_correctLength_returnsCorrectError() {
+  func testValidateCVVIncorrectLengthReturnsCorrectError() {
     Card.Scheme.allCases.forEach { scheme in
-      guard let cvvLength = scheme.cvvLength else {
-        return
-      }
-
-      let cvv = String((0..<cvvLength).map { _ in "0123456789".randomElement()! })
-      let result = subject.validate(
-        cvv: cvv,
-        cardScheme: scheme)
-
-      switch result {
-      case .success:
-        break
-      case .failure(let error):
-        XCTFail(error.localizedDescription)
-      }
-    }
-  }
-
-  func test_validateCVV_cardSchemeUnknown_incorrectLength_returnsCorrectError() {
-    let cardScheme = Card.Scheme.unknown
-
-    var result = subject.validate(
-      cvv: "12",
-      cardScheme: cardScheme)
-
-    switch result {
-    case .success:
-      XCTFail("Unexpected successful CVV validation.")
-    case .failure(let error):
-      XCTAssertEqual(error, .invalidLength)
-    }
-
-    result = subject.validate(
-      cvv: "12345",
-      cardScheme: cardScheme)
-
-    switch result {
-    case .success:
-      XCTFail("Unexpected successful CVV validation.")
-    case .failure(let error):
-      XCTAssertEqual(error, .invalidLength)
-    }
-  }
-
-
-  func test_validateCVV_incorrectLength_returnsCorrectError() {
-    // Too short
-    Card.Scheme.allCases.forEach { scheme in
-      guard let cvvLength = scheme.cvvLength else {
-        return
-      }
-
-      let cvv = String((0..<(cvvLength - 1)).map { _ in "0123456789".randomElement()! })
-      let result = subject.validate(
-        cvv: cvv,
-        cardScheme: scheme)
-
-      switch result {
-      case .success:
-        XCTFail("Unexpected successful CVV validation.")
-      case .failure(let error):
-        XCTAssertEqual(error, .invalidLength)
-      }
-    }
-
-    // Too long
-    Card.Scheme.allCases.forEach { scheme in
-      guard let cvvLength = scheme.cvvLength else {
-        return
-      }
-
-      let cvv = String((0..<(cvvLength + 1)).map { _ in "0123456789".randomElement()! })
-      let result = subject.validate(
-        cvv: cvv,
-        cardScheme: scheme)
-
-      switch result {
-      case .success:
-        XCTFail("Unexpected successful CVV validation.")
-      case .failure(let error):
-        XCTAssertEqual(error, .invalidLength)
+      for lenght in scheme.cvvLengths {
+        // Check too short
+        guard lenght > 0 else {
+          return
+        }
```

Comment:
The goal was too protect the `tooShort` test from running on a length of 0 as it would lead to a crash. I have now fixed to use an `if > 0`, to allow the `tooLong` test to run when input is 0.

Does that answer the question?

Reply:
As far i understand  that`scheme.cvvLengths` can be `[0, 3, 4]`. 
In this test case, you are looping on all predefined schemes in `Checkout SDK` that has already `0`.
According to `guard` line 57, if  the length is 0 then it should return nothing? 
i expect something like 
`func XCTFail(String, file: StaticString, line: UInt)`

Reply:
I have changed the test to no ignore `unknown` from the test and that is being tested separately. For visibility, unknown is tested as:
```swift
    func testUnknownSchemeLenghtsAndValidations() {
        let scheme = Card.Scheme.unknown
        XCTAssertEqual(scheme.cvvLengths, [0, 3, 4])
        
        let validator = CVVValidator()
        
        XCTAssertEqual(validator.validate(cvv: "", cardScheme: scheme), .success)
        XCTAssertEqual(validator.validate(cvv: "1", cardScheme: scheme), .failure(.invalidLength))
        XCTAssertEqual(validator.validate(cvv: "12", cardScheme: scheme), .failure(.invalidLength))
        XCTAssertEqual(validator.validate(cvv: "123", cardScheme: scheme), .success)
        XCTAssertEqual(validator.validate(cvv: "1234", cardScheme: scheme), .success)
        XCTAssertEqual(validator.validate(cvv: "12345", cardScheme: scheme), .failure(.invalidLength))
    }
```

In my opinion I would have done a similar test for each card scheme, but as it exists in a more generic approach I am not sure this is the right time to similarly check all other schemes. The test as it was works fine for all other schemes besides unknown.